### PR TITLE
Removed co-location groups from bulk of pipe API.

### PR DESCRIPTION
**Pull request by mwhittaker**
Recall that, before this PR, the deployer API had four methods related to weavelets and components: `RegisterReplica`, `StartComponent`, `GetComponentsToStart`, and `GetRoutingInfo`. The notion of a co-location group was baked into these APIs. A weavelet would ask for a group's routing info, not a component's, for example.

This PR removes the idea of a co-location group from the APIs and makes them component-centric instead. This is desirable because co-location groups are implementation details and should not be forced into the API. Deployers are still free to use the notion of a co-location group internally---and right now, all our deployers do---but they do not have to.

Note that this change does not eliminate co-location groups from the API entirely. There are still some message that reference groups, and the weavelet still uses the idea of a group. Future changes to the pipe API should eliminate these entirely.

### CODE REVIEW

File: internal/babysitter/babysitter.go [hunk 1]
```diff
@@ -232,85 +292,87 @@ func (b *Babysitter) StartComponent(req *protos.ComponentToStart) error {
...
-	for i, assignment := range g.routing.Val.Assignments {
-		g.routing.Val.Assignments[i] = routingAlgo(assignment, g.routing.Val.Replicas)
-	}
-
-	// Add pid.
 	g.mu.Lock()
 	defer g.mu.Unlock()
+
+	// Update addresses and pids.
+	if g.addresses[req.Address] {
```
Comment (spetrovic77) [inline internal/babysitter/babysitter.go:322]: Ditto. Let's try avoid double-locking, and update the addresses/pids here, release the lock, and then update routing.

File: weavelet.go [hunk 2]
```diff
@@ -60,8 +60,15 @@ type weavelet struct {
 	componentsByName map[string]*component       // component name -> component
 	componentsByType map[reflect.Type]*component // component type -> component
 
+	// TODO(mwhittaker): We have one client for every component. Every client
```
Comment (spetrovic77) [inline weavelet.go:63]: True that!

**Review by spetrovic77 — COMMENTED** [review]
Looks great, modulo the two-locking issue in the babysitter.

(Note that this problem really started after we changed the versioned.Versioned to acquire locks!!)

File: internal/babysitter/babysitter.go [hunk 3]
```diff
@@ -148,22 +179,41 @@ func (b *Babysitter) RegisterStatusPages(mux *http.ServeMux) {
...
+			routings:   map[string]*versioned.Versioned[*protos.RoutingInfo]{},
 		}
 		b.groups[name] = g
 	}
 	return g
 }
 
+// routing returns the RoutingInfo for the provided component.
+//
+// REQUIRES: g.mu is held.
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply internal/babysitter/babysitter.go]: Done.

File: weavelet.go (hunk 2 above)
Comment (mwhittaker) [reply weavelet.go:70]: I think I have a not-too-ugly way of fixing this that I'll do in a future PR.

File: internal/babysitter/babysitter.go (hunk 1 above)
Comment (mwhittaker) [reply internal/babysitter/babysitter.go]: Done.

File: internal/babysitter/babysitter.go [hunk 4]
```diff
@@ -346,14 +408,18 @@ func (b *Babysitter) ExportListener(req *protos.ExportListenerRequest) (*protos.
...
+// GetRoutingInfo implements the envelope.EnvelopeHandler interface.
 func (b *Babysitter) GetRoutingInfo(req *protos.GetRoutingInfo) (*protos.RoutingInfo, error) {
-	g := b.group(req.Group)
-	version := g.routing.RLock(req.Version)
-	defer g.routing.RUnlock()
-	routing := protomsg.Clone(g.routing.Val)
-	routing.Version = version
-	return routing, nil
+	g := b.group(req.Component)
+	g.mu.Lock()
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply internal/babysitter/babysitter.go]: Done.

**Comment by mwhittaker** [comment]
Thanks for the thorough review Srdjan! I made the suggested fixes in a separate commit, if you want to look at the changes only: c68d63365719a99e85a6b826954a74cdd8cfd490.

**Review by spetrovic77 — APPROVED** [review]
Looks great, thanks.

**Pull request merged**